### PR TITLE
Add multiple versions of the So2Sat dataset

### PR DESCRIPTION
**Pull request by calebrob6**
- Closes https://github.com/microsoft/torchgeo/issues/1086
- Closes the RGB request from https://github.com/microsoft/torchgeo/issues/1249 (will post an example of how you can binarize the labels with the current datamodule)
- Fixes the TODO in which we, for some reason, decided not to compute the mean and stdev of the S1 bands for So2Sat...

### CODE REVIEW

**Comment by calebrob6** [comment]
Tests are currently failing because our min version of pytorch is 1.12.0, and percentages aren't supported by `torch.utils.data.random_split` until 1.13.0. I bumped the min version because in our deprecated `torchgeo.datamodules.utils.dataset_split` we recommend using `random_split` with percentages.

File: torchgeo/datamodules/so2sat.py [hunk 1]
```diff
@@ -14,79 +16,232 @@
...
+                0.11465267892206,
+                0.15870528223797,
+                0.18053964470203,
+                0.17366821871719,
+                0.19390983961551,
+                0.15536490486611,
+                0.11057334452833,
+            ]
+        ),
+        "3_culture_10": torch.tensor(  # note: this is the same as "2"
```
Comment (adamjstewart) [inline torchgeo/datamodules/so2sat.py:91]: Could add a second line:
```python
means_per_version["3_culture_10"] = means_per_version["2"]
```

File: torchgeo/datasets/so2sat.py [hunk 2]
```diff
@@ -150,6 +183,7 @@ def __init__(
 
         Args:
             root: root directory where dataset can be found
+            version: one of "2", "3_random", "3_block", or "3_culture_10"
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:186]: Can you add a description of what `3_*` means? Either in the class or `__init__` docstring.

File: torchgeo/datamodules/so2sat.py [hunk 3]
```diff
@@ -14,79 +16,232 @@
...
+        version: str = "2",
         band_set: str = "all",
+        validation_pct: float = 0.2,
         **kwargs: Any,
     ) -> None:
         """Initialize a new So2SatDataModule instance.
 
         Args:
             batch_size: Size of each mini-batch.
+            version: One of "2" or "3_random", "3_block", or "3_culture_10".
```
Comment (adamjstewart) [inline torchgeo/datamodules/so2sat.py:219]: We don't need a `version` parameter, we can pass it directly through `**kwargs` to the dataset. See bigearthnet/oscd/seco for an example where we pull the info we need from kwargs and use it to set mean/std dev.

**Review by adamjstewart — COMMENTED** [review]
Not sure if @nilsleh @wangyi111 want to review since the dataset is from their lab.

File: torchgeo/datamodules/so2sat.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/datamodules/so2sat.py:91]: sick

File: torchgeo/datamodules/so2sat.py [hunk 4]
```diff
@@ -100,9 +209,22 @@ def setup(self, stage: str) -> None:
         Args:
             stage: Either 'fit', 'validate', 'test', or 'predict'.
         """
-        if stage in ["fit"]:
-            self.train_dataset = So2Sat(split="train", **self.kwargs)
-        if stage in ["fit", "validate"]:
-            self.val_dataset = So2Sat(split="validation", **self.kwargs)
-        if stage in ["test"]:
-            self.test_dataset = So2Sat(split="test", **self.kwargs)
+        if self.kwargs["version"] == "2":
```
Comment (adamjstewart) [inline torchgeo/datamodules/so2sat.py:212]: Should also use kwargs.get here in case this isn't provided

**Comment by adamjstewart** [comment]
Can you rebase/merge? I added EOL chars to all config files

**Comment by calebrob6** [comment]
Done

File: tests/conf/so2sat_rgb.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  num_classes: 17
+
+datamodule:
+  _target_: torchgeo.datamodules.So2SatDataModule
+  root: "tests/data/so2sat"
+  batch_size: 1
+  num_workers: 0
+  version: "3_random"
+  band_set: "rgb"
+  validation_pct: 0.5
```
Comment (adamjstewart) [inline tests/conf/so2sat_rgb.yaml:18]: Can you add an EOL char here? See #1286

File: torchgeo/datamodules/so2sat.py [hunk 6]
```diff
@@ -14,79 +16,186 @@
...
+        ),
+    }
+    stds_per_version["3_culture_10"] = stds_per_version["2"]
 
     def __init__(
         self,
         batch_size: int = 64,
         num_workers: int = 0,
         band_set: str = "all",
+        validation_pct: float = 0.2,
```
Comment (adamjstewart) [inline torchgeo/datamodules/so2sat.py:169]: In all other datamodules we call this val_split_pct, should prob be consistent.

File: torchgeo/datasets/so2sat.py [hunk 7]
```diff
@@ -24,18 +24,24 @@ class So2Sat(NonGeoDataset):
     acquired by the Sentinel-1 and Sentinel-2 remote sensing satellites, and a
     corresponding local climate zones (LCZ) label. The dataset is distributed over
     42 cities across different continents and cultural regions of the world, and comes
-    with a split into fully independent, non-overlapping training, validation,
-    and test sets.
+    with in a variety of different splits.
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:27]:
```suggestion
    with a variety of different splits.
```
Or some other grammar change

File: torchgeo/datasets/so2sat.py [hunk 8]
```diff
@@ -24,18 +24,24 @@ class So2Sat(NonGeoDataset):
...
 
-    This implementation focuses on the *2nd* version of the dataset as described in
-    the author's github repository https://github.com/zhu-xlab/So2Sat-LCZ42 and hosted
-    at https://mediatum.ub.tum.de/1483140. This version is identical to the first
-    version of the dataset but includes the test data. The splits are defined as
-    follows:
+    This implementation covers the *2nd* and *3rd* versions of the dataset as described
+    in the author's github repository https://github.com/zhu-xlab/So2Sat-LCZ42.
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:30]:
```suggestion
    in the author's github repository: https://github.com/zhu-xlab/So2Sat-LCZ42.
```

File: torchgeo/datasets/so2sat.py [hunk 9]
```diff
@@ -24,18 +24,24 @@ class So2Sat(NonGeoDataset):
...
+        train/val/test as follows:
+            * Training: 42 cities around the world
+            * Validation: western half of 10 other cities covering 10 cultural zones
+            * Testing: eastern half of the 10 other cities
+    * Version 3: A version of the dataset with 3 different train/test splits, as
+        follows:
+            * Random split: every city 80% training / 20% testing (randomly sampled)
+            * Block split: every city is split in a geospatial 80%/20%-manner
+            * Cultural 10: 10 cities from different cultural zones are held back for
+                testing purposes
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:44]: Doc formatting here is completely messed up: https://torchgeo--1283.org.readthedocs.build/en/1283/api/datasets.html#torchgeo.datasets.So2Sat

Not sure exactly how to fix this. I would suggest adding empty lines before or after outer bullet points or removing colons from the end of lines or decrease indentation of inner bullet points or make the outer bullet point lines super long (if necessary)

File: torchgeo/datasets/so2sat.py [hunk 10]
```diff
@@ -63,7 +69,8 @@ class So2Sat(NonGeoDataset):
 
     .. note::
 
-       This dataset can be automatically downloaded using the following bash script:
+       The version "2" dataset can be automatically downloaded using the following bash
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:72]:
```suggestion
       The version 2 dataset can be automatically downloaded using the following bash
```

File: torchgeo/datasets/so2sat.py [hunk 11]
```diff
@@ -74,18 +81,56 @@ class So2Sat(NonGeoDataset):
 
        or manually downloaded from https://dataserv.ub.tum.de/index.php/s/m1483140
        This download will likely take several hours.
-    """
 
-    filenames = {
-        "train": "training.h5",
-        "validation": "validation.h5",
-        "test": "testing.h5",
+       The version "3*" datasets can be downloaded using the following bash script:
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:85]:
```suggestion
       The version 3 datasets can be downloaded using the following bash script:
```
Just to make things simple

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Bunch of minor formatting changes, plus parameter rename and doc fix

File: tests/conf/so2sat_rgb.yaml (hunk 5 above)
Comment (calebrob6) [reply tests/conf/so2sat_rgb.yaml:18]: omg just turn your notifications off

Comment (adamjstewart) [reply tests/conf/so2sat_rgb.yaml:18]: I wish...

File: torchgeo/datasets/so2sat.py [hunk 12]
```diff
@@ -24,19 +24,28 @@ class So2Sat(NonGeoDataset):
...
-    at https://mediatum.ub.tum.de/1483140. This version is identical to the first
-    version of the dataset but includes the test data. The splits are defined as
-    follows:
+    This implementation covers the *2nd* and *3rd* versions of the dataset as described
+    in the author's github repository: https://github.com/zhu-xlab/So2Sat-LCZ42.
+
+    The different versions are as follows:
+
+    Version 2: This version contains imagery from 52 cities and is split into
+        train/val/test as follows:
```
Comment (adamjstewart) [inline torchgeo/datasets/so2sat.py:35]:
```suggestion
    train/val/test as follows:
```
I think it's the sudden indentation that it doesn't like.

File: tests/conf/so2sat_rgb.yaml (hunk 5 above)
Comment (adamjstewart) [inline tests/conf/so2sat_rgb.yaml:18]:
```suggestion
  val_split_pct: 0.5
```

**Review by adamjstewart — APPROVED** [review]
LGTM, assuming docs look good

**Comment by calebrob6** [comment]
Yep I checked in the last build
